### PR TITLE
codeintel: Do not use copy of scanner slice

### PR DESCRIPTION
**Pull request by efritz**
Fixes possible error condition as pointed out by @keegancsmith in https://github.com/sourcegraph/sourcegraph/pull/10990#discussion_r432080003.

### CODE REVIEW

**Comment by efritz** [comment]
Edit: There's no significant difference over three runs of the benchmark I'm using after this change.

File: cmd/precise-code-intel-worker/internal/correlation/lsif/lines/reader.go [hunk 1]
```diff
@@ -29,15 +31,21 @@ func Read(ctx context.Context, r io.Reader, unmarshal func(line []byte) (lsif.El
 	scanner.Split(bufio.ScanLines)
 	scanner.Buffer(make([]byte, LineBufferSize), LineBufferSize)
 
-	// Read the document in a separate go-routine
-	lineCh := make(chan []byte, ChannelBufferSize)
+	// Pool of buffers used to transfer copies of the scanner slice to unmarshal workers
+	pool := sync.Pool{New: func() interface{} { return new(bytes.Buffer) }}
```
Comment (keegancsmith) [inline cmd/precise-code-intel-worker/internal/correlation/lsif/lines/reader.go:35]: might as well make this a global and pre-allocate to `LineBufferSize`

File: cmd/precise-code-intel-worker/internal/correlation/lsif/lines/reader.go [hunk 2]
```diff
@@ -29,15 +31,21 @@ func Read(ctx context.Context, r io.Reader, unmarshal func(line []byte) (lsif.El
...
 		defer close(lineCh)
 
 		for scanner.Scan() {
 			if line := scanner.Bytes(); len(line) != 0 {
+				buf := pool.Get().(*bytes.Buffer)
+				_, _ = buf.Write(line)
+
 				select {
-				case lineCh <- line:
+				case lineCh <- buf:
```
Comment (keegancsmith) [inline cmd/precise-code-intel-worker/internal/correlation/lsif/lines/reader.go:48]: if you make the pool a global, you should put the buf back into it if ctx is canceled

File: cmd/precise-code-intel-worker/internal/correlation/lsif/lines/reader.go (hunk 1 above)
Comment (efritz) [reply cmd/precise-code-intel-worker/internal/correlation/lsif/lines/reader.go:35]: It's very unlikely that many buffers will become that large. It's only on a few odd edges (metadata, contains edges for large documents).

**Comment by efritz** [comment]
@keegancsmith I plan to share benchmark info between 3.16 and 3.17 as part of the release/blog post. Saving up for a big payoff 😛

**Pull request merged**